README.md
# Hackerone::Client

A limited client library for interacting with HackerOne. Currently only supports a few operations:

```ruby
client = HackerOne::Client::Api.new("github")

# POST '/reports' creates a new report
client.create_report(title: "hi", summary: "hi", impact: "string", severity_rating: :high, source: "api")

# GET '/reports' returns all reports in a given state for a program, by default :new
client.reports(state: :new)

# GET '/report/{id}' returns report data for a given report
report = client.report(id)

# PUT '/reports/{id}/assignee'
report.assign_to_user("username")
report.assign_to_group("groupname")

# POST '/reports/#{id}/activities'
report.add_comment(message, internal: false) # internal is true by default

# POST '/report/{id}/state_change change the state of a report
# `state` can be one of  new, triaged, needs-more-info, resolved, not-applicable, informative, duplicate, spam
# when marking as duplicate, you can supply the original report ID
report.state_change(:duplicate, "Your issue has been marked as X", original_report_id: 12345)

# POST '/report/{id}/add_report_reference add a "reference" e.g. internal issue number
report.add_report_reference(reference)

# Triage an issue (add a reference and set state to :triaged)
report.triage(reference)

# Set the severity on a report (rating can be :none, :low, :medium, :high or :critical)
report.update_severity(rating: :high)

# POST /reports/{id}/bounty_suggestions
report.suggest_bounty(message: "I suggest $500 with a small bonus. Report is well-written.", amount: 500, bonus_amount: 50)

# POST /reports/{id}/bounties
report.award_bounty(message: "Here's your bounty!", amount: 500, bonus_amount: 50)

# POST /reports/{id}/swags
report.award_swag(message: "Here's your T-Shirt")

# GET `/{program}/reporters` returns a list of unique reporters that have reported to your program
client.reporters

program = HackerOne::Client::Program.find("insert-program-name-here")

# returns all common responses
program.common_responses

# Updates a program's policy
program.update_policy(policy: "Please submit valid vulnerabilities")
```

## State change hooks

You can add hooks that will be called for every state change. This can be useful e.g. for ensuring that reports always get assigned or calling out to external services for specific state changes.

```ruby
# Initialization

HackerOne::Client::Report.add_state_change_hook ->(report, old_state, new_state) do
  # ...
end
```

## Usage

### Credential management

You'll need to generate an API token at `https://hackerone.com/<program>/api`.

* Click "Create API token"
* Name the token
* Click "Create"
* Copy down the value

**Set the `HACKERONE_TOKEN` and `HACKERONE_TOKEN_NAME` environment variables.**

### Program name

In order to retrieve all reports for a given program, you need to supply a default program:

```ruby
HackerOne::Client.program = "github"
```

### Risk classification

Configure the low/med/high/crit ranges for easier classification based on payouts:

```ruby
HackerOne::Client.low_range = 1..999
HackerOne::Client.medium_range = 1000...2500
HackerOne::Client.high_range = 2500...5000
HackerOne::Client.critical_range = 5000...100_000_000
```

### Configuration

In order to configure whether error handling is strict or lenient, set the `HACKERONE_CLIENT_LENIENT_MODE` variable.

Setting this variable will make the client try to absorb errors, like a malformed bounty or bonus amount. Not setting this variable will cause the client to raise errors.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/oreoshake/hackerone-client. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).


CHANGELOG.md
## [0.20.0] - 2020-10-21

[Add ability to lock a report](https://github.com/oreoshake/hackerone-client/pull/59) (@rzhade3)
[Add ability to list attachments on a report](https://github.com/oreoshake/hackerone-client/pull/58) (@rzhade3)

## [0.19.0] - 2020-05-08

[Add ability to create a report](https://github.com/oreoshake/hackerone-client/pull/57) (@rzhade3)

## [0.18.0] - 2020-03-23

[Add ability to filter reports on states other than "new"](https://github.com/oreoshake/hackerone-client/pull/54) (@rzhade3)

## [0.17.0] - 2020-03-23

[Add support for updating the severity of an issue](https://github.com/oreoshake/hackerone-client/pull/50) (@rzhade3)

## [0.16.0] - 2020-03-23

- Whooops, no change

## [0.15.0] - 2020-03-09

- [Add 'update program policy' API support](https://github.com/oreoshake/hackerone-client/pull/47) (@rzhade3)

## [0.14.2] - 2020-03-09

- [Don't lock development dependency versions](https://github.com/oreoshake/hackerone-client/pull/45) (@oreoshake)

## [0.14.1] - 2020-02-13

- [Add support for float values in API responses](https://github.com/oreoshake/hackerone-client/pull/44) (@rzhade3)

## [0.14.0] - 2019-10-28

- [Add new swag API support](https://github.com/oreoshake/hackerone-client/pull/41) (@anglinb)

## [0.13.0] - 2018-10-19

- [Fix issue with `#to_owasp` which would raise an error if an issue classification started with `CAPEC-`](https://github.com/oreoshake/hackerone-client/pull/39) (@rzhade3)

## [0.12.0] - 2017-12-07

- [Incremental activities](https://github.com/oreoshake/hackerone-client/pull/36) - iterate over actions in a program (@esjee)

## [0.11.0] - 2017-11-03

- [Add state change hooks](https://github.com/oreoshake/hackerone-client/issues/25) (@esjee)

## [0.10.0] - 2017-11-03

Oh no! I hate versions where one number is >= 10! We should figure out what's missing for a solid 1.0 release.

- `HackerOne::Client::User.find(user_id)` returns an instance of `User` (@esjee)
- Add "original report ID" field when marking things as duplicate. This invites the reporter to the original issue.

## [0.9.1] - 2017-10-24

- Misc: loosen version restrictions on activesupport

## [0.9.0] - 2017-10-09

- API: move actions from client into report (@esjee)

This is a breaking change, but this is still not a 1.0 and shouldn't be considered stable.

## [0.8.0] - 2017-09-05

- Feature: add ability to suggest and award swag, cash, and bonuses (@esjee)

## [0.7.0] - 2017-08-28

- Feature: retrieve common responses (@esjee)

## [0.6.0] - 2017-07-24

- Feature: comments (internal or not) can be added to reports

## [0.5.2] - 2017-07-19

- Bugfix: structured scopes were not being populated correctly resulting in nil results for all attributes

## [0.5.1] - 2017-06-26

- [Structure scope](https://api.hackerone.com/docs/v1#structured-scope) data added to report object

## [0.5.0] - 2017-06-23

- `report.assign_to_user` and `report.assign_to_group` (@esjee)

## [0.4.0] - 2017-04-21

- `client.reporters` to return all reporters for a given project (@esjee)
- `HackerOne::Client::Program.find(program_name)` to return information about a given program (@esjee)


lib/hackerone/client.rb
# frozen_string_literal: true

require "faraday"
require "json"
require "active_support/time"
require_relative "client/version"
require_relative "client/report"
require_relative "client/activity"
require_relative "client/program"
require_relative "client/reporter"
require_relative "client/member"
require_relative "client/user"
require_relative "client/group"
require_relative "client/structured_scope"
require_relative "client/swag"
require_relative "client/address"
require_relative "client/attachment"
require_relative "client/bounty"
require_relative "client/incremental/activities"
require "active_support/core_ext/hash"

module HackerOne
  module Client
    class NotConfiguredError < StandardError; end

    DEFAULT_LOW_RANGE = 1...999
    DEFAULT_MEDIUM_RANGE = 1000...2499
    DEFAULT_HIGH_RANGE = 2500...4999
    DEFAULT_CRITICAL_RANGE = 5000...100_000_000

    LENIENT_MODE_ENV_VARIABLE = "HACKERONE_CLIENT_LENIENT_MODE"

    REPORT_STATES = %w(
      new
      triaged
      needs-more-info
      resolved
      not-applicable
      informative
      duplicate
      spam
    )

    class << self
      ATTRS = [:low_range, :medium_range, :high_range, :critical_range].freeze
      attr_accessor :program
      attr_reader *ATTRS

      ATTRS.each do |attr|
        define_method "#{attr}=" do |value|
          raise ArgumentError, "value must be a range object" unless value.is_a?(Range)
          instance_variable_set :"@#{attr}", value
        end
      end
    end

    class Api
      def initialize(program = nil)
        @program = program
      end

      def program
        @program || HackerOne::Client.program
      end

      def reporters
        raise ArgumentError, "Program cannot be nil" unless program
        response = self.class.hackerone_api_connection.get do |req|
          req.url "programs/#{Program.find(program).id}/reporters"
        end

        data = self.class.parse_response(response)
        if data.nil?
          raise RuntimeError, "Expected data attribute in response: #{response.body}"
        end

        data.map do |reporter|
          Reporter.new(reporter)
        end
      end

      ## Returns all reports in a given state, optionally with a time bound
      #
      # program: the HackerOne program to search on (configure globally with Hackerone::Client.program=)
      # since (optional): a time bound, don't include reports earlier than +since+. Must be a DateTime object.
      # state (optional): state that a report is in, by default new
      #
      # returns all open reports or an empty array
      def reports(since: 3.days.ago, state: :new)
        raise ArgumentError, "Program cannot be nil" unless program
        raise ArgumentError, "State is invalid" unless REPORT_STATES.include?(state.to_s)

        response = self.class.hackerone_api_connection.get do |req|
          options = {
            "filter[state][]" => state,
            "filter[program][]" => program,
            "filter[created_at__gt]" => since.iso8601
          }
          req.url "reports", options
        end

        data = self.class.parse_response(response)

        data.map do |report|
          Report.new(report)
        end
      end

      ## Public: create a new report
      #
      # title: The title of the report
      # summary: Summary of the report
      # impact: Impact of the report
      # severity_rating: severity of report, must be one of https://api.hackerone.com/reference/#severity-ratings
      # source: where the report came from, i.e. API, Bugcrowd, etc.
      #
      # returns an HackerOne::Client::Report object or raises an error if
      # error during creation
      def create_report(title:, summary:, impact:, severity_rating:, source:)
        raise ArgumentError, "Program cannot be nil" unless program

        data = {
          "data": {
            "type": "report",
            "attributes": {
              "team_handle": program,
              "title": title,
              "vulnerability_information": summary,
              "impact": impact,
              "severity_rating": severity_rating,
              "source": source
            }
          }
        }
        Report.new(post("reports", data))
      end

      ## Public: retrieve a report
      #
      # id: the ID of a specific report
      #
      # returns an HackerOne::Client::Report object or raises an error if
      # no report is found.
      def report(id)
        Report.new(get("reports/#{id}"))
      end

      private
      def post(endpoint, body)
        response = with_retry do
          self.class.hackerone_api_connection.post do |req|
            req.headers["Content-Type"] = "application/json"
            req.body = body.to_json
            req.url endpoint
          end
        end

        self.class.parse_response(response)
      end

      def get(endpoint, params = nil)
        response = with_retry do
          self.class.hackerone_api_connection.get do |req|
            req.headers["Content-Type"] = "application/json"
            req.params = params || {}
            req.url endpoint
          end
        end

        self.class.parse_response(response)
      end

      def self.parse_response(response, extract_data: true)
        if response.status.to_s.start_with?("4")
          raise ArgumentError, "API called failed, probably your fault: #{response.body}"
        elsif response.status.to_s.start_with?("5")
          raise RuntimeError, "API called failed, probably their fault: #{response.body}"
        elsif response.success?
          response_body_json = JSON.parse(response.body, symbolize_names: true)
          if extract_data && response_body_json.key?(:data)
            response_body_json[:data]
          else
            response_body_json
          end
        else
          raise RuntimeError, "Not sure what to do here: #{response.body}"
        end
      end

      def self.hackerone_api_connection
        unless ENV["HACKERONE_TOKEN_NAME"] && ENV["HACKERONE_TOKEN"]
          raise NotConfiguredError, "HACKERONE_TOKEN_NAME HACKERONE_TOKEN environment variables must be set"
        end

        @connection ||= Faraday.new(url: "https://api.hackerone.com/v1") do |faraday|
          faraday.basic_auth(ENV["HACKERONE_TOKEN_NAME"], ENV["HACKERONE_TOKEN"])
          faraday.adapter Faraday.default_adapter
        end
      end

      def with_retry(attempts = 3, &block)
        attempts_remaining = attempts

        begin
          yield
        rescue StandardError
          if attempts_remaining > 0
            attempts_remaining -= 1
            sleep (attempts - attempts_remaining)
            retry
          else
            raise
          end
        end
      end
    end
  end
end
